### PR TITLE
ENH: Add histogram_bin_edges function and test

### PR DESCRIPTION
**Pull request by kirit93**
PR for #10183 
@eric-wieser CircleCI seems to fail due to an Extension-Error. Any idea what might have caused this?

### CODE REVIEW

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: This is a good start. You should take a look at [HOWTO_DOCUMENT](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections) - in particular, a parameters section and a couple examples would go a long way

File: numpy/lib/tests/test_histograms.py
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: Aligning at equals like this isn't recommended by PEP8, so we tend not to do it in numpy. Otherwise these tests look reasonable. Maybe add one for a string `bins` argument too?

File: numpy/lib/tests/test_histograms.py
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: You should all this to `__all__` in `histograms.py` too.

**Comment by eric-wieser** [comment]
Don't worry about circle CI, it's been failing for a while. @stefanv, can you take a look at what's going on with that at some point?

File: numpy/lib/tests/test_histograms.py
Comment (kirit93) [reply numpy/lib/tests/test_histograms.py]: Done.

File: numpy/lib/tests/test_histograms.py
Comment (kirit93) [reply numpy/lib/tests/test_histograms.py]: Okay, will keep this in mind henceforth. Added another test.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: If you put backticks around the word `histogram`, it will create a link to the function in the docs

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: I wonder whether we should remove this help from `np.histogram`, and leave it here exclusively?

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Probably worth noting that this is currently ignored, but maybe be needed by future bin estimators.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Same remark about alignment here

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: This looks like what you get when you `print()` the result - which isn't what will actually appear if you type this into an interactive python terminal. 

You should show `arr` being created here, and add an `Example` heading too.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Sorry, missed this. Resolved now.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Fixed this.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: It is a bit redundant to have the different types of bins documented twice. If you'd like I'll remove it form ```np.histogram``` and then push the commit with all the fixes.

Comment (eric-wieser) [reply numpy/lib/histograms.py]: Let's see if anyone else has anything to say about that first

Comment (kirit93) [reply numpy/lib/histograms.py]: Sure thing. I'll push the latest commit leaving the doc for ```np.histogram``` intact. In case you want it changed later, I'll modify the docstring.

**Comment by stefanv** [comment]
@eric-wieser It is being addressed in https://github.com/numpy/numpy/issues/10595

File: numpy/lib/tests/test_histograms.py
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: nit: PEP8 wants two newlines between one class and the next - so one more here

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Alignment looks off here - I think you're missing 4 spaces. I also think a much shorter example with a result that fits on one line would be clearer

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: This isn't how arrays are printed any more in 1.14.

Still needs a line of code showing how `arr` was created.

File: numpy/lib/histograms.py
Comment (chunweiyuan) [reply numpy/lib/histograms.py]: I face a somewhat similar dilemma with `np.quantile` and `np.percentile` (and `np.nanquantile`,  `np.nanpercentile`) in #9211.  I think in #9213 there's a comment about `just make the usual docstring`, so I went the verbose way.

File: numpy/lib/histograms.py
Comment (chunweiyuan) [reply numpy/lib/histograms.py]: Could we specify whether these weights can be negative (I assume they can't due to normalization)?  Given that we have the `weights` arg in various functions now, and that they have different nuances, it might be good practice to be a little more detailed here.

Comment (eric-wieser) [reply numpy/lib/histograms.py]: @chunweiyuan: Well, since they're unused they can be literally anything :). But your comment is valid when applied to `np.histogram`

**Comment by charris** [comment]
Circleci should be fixed in master if you want to rebase.

File: numpy/lib/tests/test_histograms.py
Comment (kirit93) [reply numpy/lib/tests/test_histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: I changed the doctstring to remove the line on ```density```. Since it isn't an argument for ```histogram_bin_edges``` edges it doesn't make sense in the docstring.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Could do with a see also section referring to histogram (and one in the other direction too). That should be in HOWTO_DOCUMENT as well.

File: doc/release/1.15.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: I think this is available directly as `np.histogram_bin_edges`

File: doc/release/1.15.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: Typo

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Maybe use `(arr, bins=2)` for this one so it has a different output? Also because an example without `range` would be good.

File: numpy/lib/histograms.py
Comment (eric-wieser) [reply numpy/lib/histograms.py]: Let's leave the duplication, and then we can propose removing it in a later PR

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Repeating the earlier remark - we should note that this is currently not used by any of the bin estimators.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Should be "may be" instead of "maybe", I think - only the former is a verb.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: nit: extra newline needed above the `def` for PEP8

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

**Comment by kirit93** [comment]
@eric-wieser, do any more changes need to be made here or is it ready to merge?

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: This example should probably come last, since it's the least useful. Maybe preface it with _"For consistency with `histogram`, an array of pre-computed bins is passed through unmodified"_, so the user knows why the result is that

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Should prefix with `np.`

Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (kirit93) [reply numpy/lib/histograms.py]: Done.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: An example showing how to compute matching histograms would be nice, since that inspired this function in the first place:

```
group_id = np.array([0, 1, 1, 0, 1, 1, 0, 1, 1])
shared_bins = np.histogram_bin_edges(arr)
hist_0, _ = np.histogram(arr[group_id == 0], bins=shared_bins)
hist_1, _ = np.histogram(arr[group_id == 1], bins=shared_bins)
```

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: No need for the `#` here. The examples section can contain plain text, as long as you separate it from the `>>>` examples with newlines

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Old-style output again. Are you aware of `runtests.py -i` which allows you to run these commands and copy their output?

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: I suspect this needs a newline between the text and the `>>>` to render correctly - and same in the other two sentences above

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: `>>>` should not be here, this is output

**Review by eric-wieser — COMMENTED** [review]
I'll fix up the formatting myself

**Comment by eric-wieser** [comment]
Rather than going back and forth on docs here, I've just added a commit with some suggested changes - feel free to further modify it if you don't like what I've changed

**Review by eric-wieser — APPROVED** [review]
LGTM pending mailing-list approval. I can write a message out if you don't want to.

**Comment by kirit93** [comment]
Thanks, @eric-wieser, I can write out a message to the mailing list.

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: This `histogram` should be `histogram_bin_edges`

File: numpy/lib/histograms.py
Comment (eric-wieser) [inline numpy/lib/histograms.py]: I don't think ``(length(hist)+1)`` makes any sense here

**Comment by charris** [comment]
@kirit93 I don't see any message to the mailing list.

**Comment by kirit93** [comment]
@charris I requested to join the mailing list but I still haven't gotten a response. I tried posting a message twice with no luck.

**Comment by charris** [comment]
@kirit93 Hmm, where did you try to sign up?

**Comment by charris** [comment]
The mailing list should be here: https://mail.python.org/mailman/listinfo/numpy-discussion .

**Comment by charris** [comment]
The mailing list response was positive, so this is good to go.

**Comment by eric-wieser** [comment]
Thanks @kirit93!

File: doc/release/1.15.0-notes.rst
Comment (madphysicist) [inline doc/release/1.15.0-notes.rst]: "...without calculating the histogram" sounds better (for a later PR).

File: numpy/lib/histograms.py
Comment (madphysicist) [reply numpy/lib/histograms.py]: @eric-wieser I think it is a good idea to remove the duplication. Would it make sense to have a separate documentation page for histogram bin estimation? This function and histogram could link to it.

**Comment by madphysicist** [comment]
Sorry for being so late to this party. Is there any reason not to replace `_get_bin_edges` with this function entirely?

File: numpy/lib/histograms.py
Comment (eric-wieser) [reply numpy/lib/histograms.py]: Already removed in #10755

**Comment by eric-wieser** [comment]
Because internal optimization formats change, and we don't want downstream users depending on them.

Potentially in future we could do something like:

```
class LinSpaceArray(np.ndarray):
    start: float64
    stop: float64
    n: intp
```
And return that instead of a raw array

**Comment by madphysicist** [comment]
I like the sneakiness. What about the docs? Do you think it is worth creating a separate page for the histogram estimators? It would make the documentation of `histogram` and this function look much cleaner and more consistent, if nothing else.

**Comment by eric-wieser** [comment]
> I like the sneakiness.

Sadly it's a little more involved since it needs [a WeakKeyIdDictionary, which is rather tricky to write](https://mail.python.org/pipermail/python-ideas/2014-December/030573.html)

> Do you think it is worth creating a separate page

I'm not sure. One thing I don't like is how the estimators have their own docstrings, which do not appear in the online docs anywhere.

**Comment by madphysicist** [comment]
@eric-wieser. I'll set up a PR for that pretty soon, and you can tell me if you don't like it for some reason :)

**Comment by eric-wieser** [comment]
Neither of the two approaches I suggest above will work:
```python
a = histogram_bin_edges(a,  bins=10)
a[4]  = 10
histogram(a)  # better have cleared our cache!
```

**Pull request merged**
